### ISSUE TITLE
Query limit method cannot be combined with both select and order_by methods

### ISSUE THREAD

**Issue by healiseu** [josiahcarlson/rom#126]
Hi, I have written a test on OneToMany relationship case. You can download my Jupyter Notebook file ( ROM - Testing One to Many Relationship.ipynb) [here][1]. JSON Datasets are also included in the same folder.

I reached a point in execution (see **Fetch top 10 items with the highest price with a joined projection from Company and Item Entity Types** section) where I need to sort 10 Items by price. It turns out that you cannot use a select-limit-order_by combination. The query is executed but the result set is not limited to 10 instances. Is that a bug or limited functionality ? Do I miss something here ?

I also want to ask if the present functionality of ROM is limited to single attribute ordering ?

[1]: https://notebooks.azure.com/athanassios/libraries/redisrom
[2]: https://redisrom-athanassios.notebooks.azure.com/nb/notebooks/ROM%20-%20Testing%20One%20to%20Many%20Relationship.ipynb

**Comment by josiahcarlson** [comment]
> The query is executed but the result set is not limited to 10 instances. Is that a bug or limited functionality ? Do I miss something here ?

You didn't miss anything, I found the bug. There's a missing `remaining -= 1` in a method in my stuff. I've added a test and fixed the bug. Will be releasing a fix soon.

> I also want to ask if the present functionality of ROM is limited to single attribute ordering ?

Yes and no. The problem is that we're fundamentally limited to a single 64 bit float for ordering using built-in structures. So if your ordering requires comparing something that can't be packed / represented in a 64 bit float, nothing in rom will be able to order it properly.

Also, there isn't a built-in mechanism for multi-column `order_by()` calls, partly because to do it right requires packing columns into an output ZSET, which are limited in precision and magnitude. I've got something that works reasonably well on integer column data of limited magnitude, and there are several things that could be done to make it more reasonable for floating point data, but we're still limited by a 64 bit float:
https://gist.github.com/josiahcarlson/7126873

**Comment by josiahcarlson** [comment]
Almost forgot: with the `keygen2=` argument to a column, you can pre-compute an ordering if you'd like, and explicitly create an index with that ordering.

```
def combine_fields(colx, coly):
    # say that we want to order by (colx, coly)
    # and we know -10000 < coly < 10000
    return colx * 10000 + coly

def compose_index(attr, dict):
        return {'': combine_fields(dict['colx'], dict['coly'])}

class IndexOnly(rom.Integer):
    _allowed = ()

class TestModel(rom.Model):
    custom_order = IndexOnly(index=True, keygen2=compose_index)
    ...

# used:
TestModel(colx=..., coly=...).save()
TestModel.query.order_by('custom_order')
```

This isn't useful for ad-hoc queries, but it is useful when you know in advance you need to do these exact queries all the time.

**Comment by josiahcarlson** [comment]
Looking more into the source of your data for #125 made me see that your source of data was using the Redis `SORT` command to order data at runtime. FWIW, rom doesn't currently use the Redis `SORT` command for operations, even if in some cases it might be more efficient in space / computation. That's mostly due to trying to stick with sets / sorted sets for consistency.

**Comment by healiseu** [comment]
Hi Josiah, I want to define a composite index based on a class with three Integer attributes (dim3, dim2, dim1) and use it to order instances of a class. 
Let's say the range of these dimensions are:
dim3  in [1 .. 4], 
dim2  in [1 .. 10000]
dim3  in [1 .. 10^9]
I understand that I have to use IndexOnly and keygen2 but I am confused on how exactly to define and use them, you also mentioned that you made some changes. Based on these specifications, can I see a complete code example with a query call to try out. I am using ROM 0.42.2

**Comment by josiahcarlson** [comment]
Code:

```
import rom

def combine(dim3, dim2, dim1):
    return (10**13 * dim3) + (10**9 * dim2) + dim1

class Demo(rom.Model):
    dim3  = rom.Integer(required=True, index=True)
    dim2  = rom.Integer(required=True, index=True)
    dim1  = rom.Integer(required=True, index=True)
    combo = rom.Integer(unique=True, index=True)

    val  = rom.Float(required=True)

    unique_together = [('dim3', 'dim2', 'val')]

    def _before_insert(self):
        self.combo = combine(self.dim3, self.dim2, self.dim1)

    _before_update = _before_insert
```

Used:
```
In [1]: import rom
   ...: 
   ...: def combine(dim3, dim2, dim1):
   ...:     return (10**13 * dim3) + (10**9 * dim2) + dim1
   ...: 
   ...: class Demo(rom.Model):
   ...:     dim3  = rom.Integer(required=True, index=True)
   ...:     dim2  = rom.Integer(required=True, index=True)
   ...:     dim1  = rom.Integer(required=True, index=True)
   ...:     combo = rom.Integer(unique=True, index=True)
   ...: 
   ...:     val  = rom.Float(required=True)
   ...: 
   ...:     unique_together = [('dim3', 'dim2', 'val')]
   ...: 
   ...:     def _before_insert(self):
   ...:         self.combo = combine(self.dim3, self.dim2, self.dim1)
   ...: 
   ...:     _before_update = _before_insert
   ...:     

In [2]: rom.util.use_null_session()

In [3]: Demo(dim3=1, dim2=2, dim1=5, val=6.7)
Out[3]: <__main__.Demo at 0x7f4dbeb47710>

In [4]: Out[3].save()
Out[4]: 6

In [5]: Demo.query.filter(combo=combine(1, 2, 5)).all()
Out[5]: [<__main__.Demo at 0x7f4dbd0709e8>]
```

**Comment by healiseu** [comment]
:+1: thanks for the code, I am surprised that you did not use `IndexOnly` and/or `keygen2` in your implementation. Is it faster/better/easier this way and why ? I am also trying to understand where you use IndexOnly and keygen2, I suppose one could also write equivalent code with IndexOnly and keygen2.

**Comment by josiahcarlson** [comment]
`IndexOnly()` doesn't currently work with `unique=True`, which is undocumented. But you also wanted a unique constraint, so I thought it easier to just make it all explicit and use a column to store the computed data. That way it works the same with `Demo.query.filter(combo=combine(1, 2, 5)).all()` and `Demo.get_by(combo=combine(1, 2, 5))`:

```
In [45]: Demo.get_by(combo=combine(1,2,5))
Out[45]: <__main__.Demo at 0x7f5c1a1c4c18>
```

With `_before_insert` and `_before_update` magic, the index column is recomputed on `.save()`, and everything else works the same.

I could have used the `IndexOnly()` column, then added another `unique_together` entry of `('dim3', 'dim2', 'dim1')`. That might be *slightly* smaller in terms of memory use, because the computed data isn't stored, except in the indexes. That one would look like:
```
import rom

def combine(dim3, dim2, dim1):
    return (10**13 * dim3) + (10**9 * dim2) + dim1

def combine_keygen(attr, dict):
    return {'': combine(dict['dim3'], dict['dim2'], dict['dim1'])}

class Demo(rom.Model):
    dim3  = rom.Integer(required=True, index=True)
    dim2  = rom.Integer(required=True, index=True)
    dim1  = rom.Integer(required=True, index=True)
    combo = rom.IndexOnly(keygen2=combine_keygen)

    val  = rom.Float(required=True)

    unique_together = [
        ('dim3', 'dim2', 'val'),
        ('dim3', 'dim2', 'dim1'),
    ]
```

Executed:
```
In [5]: a = Demo(dim3=1, dim2=2, dim1=5, val=6.7)
   ...: a.save()
   ...: 
   ...: Demo.query.filter(combo=combine(1, 2, 5)).all()
   ...: 
Out[5]: [<__main__.Demo at 0x7f3cf43048d0>]
```

Only difference is the output of `Demo.get_by()`:

```
In [7]: Demo.get_by(combo=combine(1, 2, 5))
Out[7]: [<__main__.Demo at 0x7f3cf43048d0>]
```

... because it's using the numeric index for `get_by()` instead of the unique index.

**Comment by healiseu** [comment]
Smashing, I think that covers me completely at the moment. Thanks for the explanations and the code that demonstrates many features in ROM that I did not know exactly how to use such as multiple unique constrains, IndexOnly and keygen2. I am pretty sure other ROM users will find the example covered here very useful.

**Comment by healiseu** [comment]
Oops there is a problem passing `keygen2` parameter in `IndexOnly`. 
Is this fixed on ROM 0.42.2. ?

``` Python
File "<ipython-input-2-9a3f9a940861>", line 25, in <module>
    class Demo(Model):
  File "<ipython-input-2-9a3f9a940861>", line 29, in Demo
    combo = IndexOnly(keygen2=composite_index)
TypeError: __init__() got an unexpected keyword argument 'keygen2'
```

**Comment by josiahcarlson** [comment]
Forgot that the updated `keygen2` argument was in a pending release. Please see 0.42.3, released moments ago.
